Source/WebCore/css/StyleColor.h

```diff
@@ -31,31 +31,86 @@
 
 #pragma once
 
-#include "CSSValueKeywords.h"
 #include "Color.h"
+#include "ColorInterpolationMethod.h"
+#include "CSSValueKeywords.h"
+#include <vector>
 #include <wtf/OptionSet.h>
 
 namespace WebCore {
 
+class ColorMixComponent;
+class CSSPrimitiveValue;
+class StyleColor;
+
 enum class StyleColorOptions : uint8_t {
     ForVisitedLink = 1 << 0,
     UseSystemAppearance = 1 << 1,
     UseDarkAppearance = 1 << 2,
     UseElevatedUserInterfaceLevel = 1 << 3
 };
 
-struct StyleColor {
+struct CurrentColor {
+    bool operator==(CurrentColor const&) const = default;
+    static StyleColor make();
+};
+
+struct SystemColor {
+    bool operator==(SystemColor const&) const = default;
+};
+struct ColorContext {
+    std::optional<Color> currentColor = std::nullopt;
+};
+
+using ColorKind = std::variant<Color, CurrentColor, SystemColor>;
+
+class StyleColor {
+public:
     static Color colorFromKeyword(CSSValueID, OptionSet<StyleColorOptions>);
+    static Color colorFromAbsoluteKeyword(CSSValueID);
     static bool isAbsoluteColorKeyword(CSSValueID);
+    static bool isCurrentColorKeyword(CSSValueID);
+    static bool isCurrentColor(CSSPrimitiveValue const&);
     WEBCORE_EXPORT static bool isSystemColorKeyword(CSSValueID);
 
     enum class CSSColorType : uint8_t {
         Absolute = 1 << 0,
         Current = 1 << 1,
         System = 1 << 2,
     };
+
     // https://drafts.csswg.org/css-color-4/#typedef-color
     static bool isColorKeyword(CSSValueID, OptionSet<CSSColorType> = { CSSColorType::Absolute, CSSColorType::Current, CSSColorType::System });
+
+    StyleColor();
+    StyleColor(Color const&);
+    StyleColor(ColorKind const&);
+    StyleColor(StyleColor const&) = default;
+    StyleColor& operator=(StyleColor const&) = default;
+    bool operator==(StyleColor const&) const = default;
+
+    StyleColor(SRGBA<uint8_t> const&);
+
+    bool isCurrentColor() const;
+    bool isSystemColor() const;
+    bool isAbsoluteColor() const;
+    bool isColorMix() const;
+    Color const& absoluteColor() const;
+    void setInheritedColor(const Color&);
+    std::optional<Color> inheritedColor() const;
+
+    WEBCORE_EXPORT Color resolveColor(ColorContext = { }) const;
+
+    friend WTF::TextStream& operator<<(WTF::TextStream&, StyleColor const&);
+    String debugDescription() const;
+
+private:
+    ColorKind m_color;
+    // The color property may depends on its inherited value at runtime.
+    std::optional<Color> m_inheritedColor;
```

Comment:
There shouldn't be any need for this field. The colors are resolved at use time. You can see there is nothing like this field at the moment and we manage ok. How would we know 

As said before, the first patch should just introduce StyleColor as a direct replacement for Color in RenderStyle, encapsulating a color value and 'current color' state. No new functionality should be included, this is complex enough change in itself. Internally it can even keep using null Color to signify 'current color' so there is no memory or performance concerns. In future patches we can look into what is the best way to implement new functionality on top.

Reply:
This is a leftover from the initial patch (with the color-mix feature) indeed.